<?xml version="1.0"?>
<!-- This launch file is used to produce human joint angle -->
  <launch>
    <!-- These lines are used to define the initializeable arguments and parameters-->
    <!-- There are several important arguments described as follows-->
    <!-- 1. Argument "wrist_on" is set to "1" when wrist sensor is used or "0" when wrist sensor is disabled.-->
    <!-- 2. Argument "index" is set to the human index selected in the Kinect nodes. For example, if the Kinect detecst the human's elbow as "left_human_2", then we set argument "index" to be "2".-->
    <!-- Important: These 2 arguments normally do not need to be initialized since they are assigned already in launch_ergonomics.launch. -->
    <arg name="index" default="1"/>
    <arg name="wrist_on" default="1"/>


      <!-- The code frame_listener.py is used to make transformation from one frame to another. In this case, frame "left_shoulder" is used as a reference for all other frames.-->
      <!-- Transform from "left_elbow" to "left_shoulder"-->
      <node pkg="kcl_ergonomics" type="frame_listener.py" name="node_elbow_shoulder" respawn="false" output="screen">
      	<param name="frame" type="string" value="left_elbow" />
      	<param name="ref" type="string" value="left_shoulder" />
      	<param name="publish_to" type="string" value="left_elbow_shoulder"/>
      	<param name="ind" type="int" value="$(arg index)"/>
      </node>

      <!-- Transform from "left_hand" to "left_shoulder"-->
      <node pkg="kcl_ergonomics" type="frame_listener.py" name="node_hand_shoulder" respawn="false">
      	<param name="frame" type="string" value="left_hand" />
      	<param name="ref" type="string" value="left_shoulder" />
      	<param name="publish_to" type="string" value="left_hand_shoulder"/>
      	<param name="ind" type="int" value="$(arg index)"/>
      </node>

      <!-- Transform from "neck" to "left_shoulder"-->
      <node pkg="kcl_ergonomics" type="frame_listener.py" name="node_neck_shoulder" respawn="false">
      	<param name="frame" type="string" value="neck" />
      	<param name="ref" type="string" value="left_shoulder" />
      	<param name="publish_to" type="string" value="left_neck_shoulder"/>
      	<param name="ind" type="int" value="$(arg index)"/>
      </node>

      <!-- Transform from "torso" to "left_shoulder"-->
      <node pkg="kcl_ergonomics" type="frame_listener.py" name="node_torso_shoulder" respawn="false">
      	<param name="frame" type="string" value="torso" />
      	<param name="ref" type="string" value="left_shoulder" />
      	<param name="publish_to" type="string" value="left_torso_shoulder"/>
      	<param name="ind" type="int" value="$(arg index)"/>
      </node>

      <!-- Transform from "neck" to "left_shoulder"-->
      <node pkg="kcl_ergonomics" type="frame_listener.py" name="node_left_right_shoulder" respawn="false">
      	<param name="frame" type="string" value="neck" />
      	<param name="ref" type="string" value="left_shoulder" />
      	<param name="publish_to" type="string" value="left_right_shoulder"/>
      	<param name="ind" type="int" value="$(arg index)"/>
      </node>

      <!-- This node is used to generate a sign for the human's joint angle. The camera NEEDS TO BE PUT IN THE RIGHT SIDE OF THE HUMAN.-->
      <node pkg="kcl_ergonomics" type="human_position.py" name="human_position" respawn="false" output="screen">
        <param name="ind" type="int" value="$(arg index)"/>
      </node>

      <!-- This node is used to generate the human's joint angle from the frame information.-->
      <node pkg="kcl_ergonomics" type="human_angle.py" name="human_angle" output="screen">
        <param name="wrist_active" type="int" value="$(arg wrist_on)" />
      </node>
  </launch>
